### PR TITLE
[RSDK-9653, RSDK-9673] - Make resilient to unknown and changing frame sizes

### PR DESCRIPTION
**Pull request by seanavery**
## Description

This addresses crashes in the encoder downstream from mismatches between the actual frames received from the source camera and the encoder/segmenter parameters. We now defer initialization of the encoder and segmenter until we receive a frame and can reinitialize when the dimensions change. If the Image endpoint of the source camera changes we will spin off a new initialization flow to handle the new size.

Beforehand you can reliably repro a crash with wildly off `cam_props`. For instance, the following parameters with a 640x480 source camera.

```
        "cam_props": {
          "height": 10,
          "framerate": 20,
          "width": 10000
        }
        
```

Will show the following segfault trace:

```
\_ Assertion ((src_linesize) >= 0 ? (src_linesize) : (-(src_linesize))) >= bytewidth failed at libavutil/imgutils.c:350
2025-01-10T17:19:11.749Z        ERROR   rdk.modmanager.viam_video-store.viam_video-store.StdErr pexec/managed_process.go:277
\_ SIGABRT: abort
2025-01-10T17:19:11.750Z        ERROR   rdk.modmanager.viam_video-store.viam_video-store.StdErr pexec/managed_process.go:277
\_ PC=0x7fbc044e6c m=9 sigcode=18446744073709551610  
2025-01-10T17:19:11.750Z        ERROR   rdk.modmanager.viam_video-store.viam_video-store.StdErr pexec/managed_process.go:277
\_ signal arrived during cgo execution
2025-01-10T17:19:11.750Z        ERROR   rdk.modmanager.viam_video-store.viam_video-store.StdErr pexec/managed_process.go:277
\_

....

```

## Tests
- Adjusted config tests for new setup without cam_props and optoinal framerate property.
- Manually tested a changing source RTSP camera.
  - Verify that current segment ends on size change trigger ✅ 
  - Verify that next segment at adjusted size is playable ✅ 
  - Verify multiple size change events work ✅ 
  
  
  ```
  {
  "components": [
    {
      "name": "video-store",
      "model": "viam:video:storage",
      "attributes": {
        "sync": "data_manager-1",
        "storage": {
          "segment_seconds": 10,
          "upload_path": "/home/viam/.viam/video-upload-test",
          "size_gb": 10
        },
        "cam_props": {
          "width": 10000,
          "height": 10,
          "framerate": 20
        },
        "video": {
          "preset": "ultrafast"
        },
        "camera": "rtsp-cam-1"
      },
      "depends_on": [
        "rtsp-cam-1",
        "data_manager-1"
      ],
      "namespace": "rdk",
      "type": "camera"
    },
    {
      "name": "rtsp-cam-1",
      "model": "viam:viamrtsp:rtsp",
      "attributes": {
        "rtsp_address": "rtsp://admin:GOSTadm1n@10.1.12.67:554/cam/realmonitor?channel=2&subtype=0&unicast=true&proto=Onvif",
        "rtp_passthrough": true
      },
      "namespace": "rdk",
      "type": "camera"
    }
  ],
  "services": [
    {
      "name": "data_manager-1",
      "attributes": {
        "sync_interval_mins": 0.1,
        "capture_dir": "",
        "tags": [],
        "additional_sync_paths": [],
        "capture_disabled": true
      },
      "namespace": "rdk",
      "type": "data_manager"
    }
  ],
  "modules": [
    {
      "type": "registry",
      "name": "viam_viamrtsp",
      "module_id": "viam:viamrtsp",
      "version": "0.2.0-rc4"
    },
    {
      "executable_path": "/home/viam/video-store-no-cam-props-review",
      "name": "viam_video-store",
      "type": "local"
    }
  ]
}
```

### CODE REVIEW

**Comment by nicksanford** [comment]
> ## Description
> This addresses crashes in the encoder downstream from mismatches between the actual frames received from the source camera and the encoder/segmenter parameters. We now defer initialization of the encoder and segmenter until we receive a frame and can reinitialize when the dimensions change. If the Image endpoint of the source camera changes we will spin off a new initialization flow to handle the new size.
> 
> Beforehand you can reliably repro a crash with wildly off `cam_props`. For instance, the following parameters with a 640x480 source camera.
> 
> ```
>         "cam_props": {
>           "height": 10,
>           "framerate": 20,
>           "width": 10000
>         }
>         
> ```
> 
> Will show the following segfault trace:
> 
> ```
> \_ Assertion ((src_linesize) >= 0 ? (src_linesize) : (-(src_linesize))) >= bytewidth failed at libavutil/imgutils.c:350
> 2025-01-10T17:19:11.749Z        ERROR   rdk.modmanager.viam_video-store.viam_video-store.StdErr pexec/managed_process.go:277
> \_ SIGABRT: abort
> 2025-01-10T17:19:11.750Z        ERROR   rdk.modmanager.viam_video-store.viam_video-store.StdErr pexec/managed_process.go:277
> \_ PC=0x7fbc044e6c m=9 sigcode=18446744073709551610  
> 2025-01-10T17:19:11.750Z        ERROR   rdk.modmanager.viam_video-store.viam_video-store.StdErr pexec/managed_process.go:277
> \_ signal arrived during cgo execution
> 2025-01-10T17:19:11.750Z        ERROR   rdk.modmanager.viam_video-store.viam_video-store.StdErr pexec/managed_process.go:277
> \_
> 
> ....
> ```
> 
> ## Tests
> * Adjusted config tests for new setup without cam_props and optoinal framerate property.
> * Manually tested a changing source RTSP camera.
>   
>   * Verify that current segment ends on size change trigger ✅
>   * Verify that next segment at adjusted size is playable ✅
>   * Verify multiple size change events work ✅

@seanavery 
1. Can you please add the configuration you used to test these things?
2. Can you please test whether or not there is a significant performance cost to this change?

File: cam/cam.go [hunk 1]
```diff
@@ -164,39 +158,16 @@ func newvideostore(
...
-		}
-	}
-	if newConf.Properties.Width == 0 && newConf.Properties.Height == 0 {
-		return nil, fmt.Errorf("failed to get source camera width and height after %d attempts", numFetchFrameAttempts)
-	}
-
-	if newConf.Properties.Framerate == 0 {
-		newConf.Properties.Framerate = defaultFramerate
+	if newConf.Framerate != 0 {
+		vs.framerate = newConf.Framerate
```
Comment (randhid) [inline cam/cam.go:162]: Please add a negative number check in validate.

**Review by randhid — CHANGES_REQUESTED** [review]
Negative numbers into those time.Durations will crash everything - please add a validation check and a test ensuring we're not passing in negative framerates.

File: cam/encoder.go [hunk 2]
```diff
@@ -27,91 +27,125 @@ type encoder struct {
...
 // PTS is calculated based on the frame count and source framerate.
 // If the polling loop is not running at the source framerate, the
 // PTS will lag behind actual run time.
-func (e *encoder) encode(frame image.Image) ([]byte, int64, int64, error) {
+func (e *encoder) encode(frame image.Image) ([]byte, int64, int64, bool, error) {
+	reinit := false
+	dy, dx := frame.Bounds().Dy(), frame.Bounds().Dx()
+	if e.codecCtx == nil || dy != int(e.codecCtx.height) || dx != int(e.codecCtx.width) {
+		e.logger.Infof("Initializing encoder with frame dimensions %dx%d", dx, dy)
+		err := e.initialize(dx, dy)
```
Comment (nicksanford) [inline cam/encoder.go:135]: can we please change `initialize` to have no side effects if it fails?

File: cam/encoder.go [hunk 3]
```diff
@@ -27,91 +27,125 @@ type encoder struct {
...
 }
 
 // encode encodes the given frame and returns the encoded data
 // in bytes along with the PTS and DTS timestamps.
 // PTS is calculated based on the frame count and source framerate.
 // If the polling loop is not running at the source framerate, the
 // PTS will lag behind actual run time.
-func (e *encoder) encode(frame image.Image) ([]byte, int64, int64, error) {
+func (e *encoder) encode(frame image.Image) ([]byte, int64, int64, bool, error) {
+	reinit := false
```
Comment (nicksanford) [inline cam/encoder.go:131]:
```suggestion
	frameDimentionsChanged := false
```

File: cam/encoder.go [hunk 4]
```diff
@@ -27,91 +27,125 @@ type encoder struct {
...
+	return nil
 }
 
 // encode encodes the given frame and returns the encoded data
 // in bytes along with the PTS and DTS timestamps.
 // PTS is calculated based on the frame count and source framerate.
 // If the polling loop is not running at the source framerate, the
 // PTS will lag behind actual run time.
-func (e *encoder) encode(frame image.Image) ([]byte, int64, int64, error) {
+func (e *encoder) encode(frame image.Image) ([]byte, int64, int64, bool, error) {
```
Comment (nicksanford) [inline cam/encoder.go:130]: make `[]byte, int64, int64, bool,` into a `(struct, err)`

File: Makefile [hunk 5]
```diff
@@ -43,7 +43,7 @@ FFMPEG_OPTS ?= --prefix=$(FFMPEG_BUILD) \
 
 CGO_LDFLAGS := -L$(FFMPEG_BUILD)/lib -lavcodec -lavutil -lavformat -lz
 ifeq ($(SOURCE_OS),linux)
-	CGO_LDFLAGS += -l:libjpeg.a -l:libx264.a
+	CGO_LDFLAGS += -l:libx264.a
```
[the remark this answers is not in the document]
Comment (seanavery) [reply Makefile:46]: Nothing I should have flagged this. I believe we are getting this error due to a change in ubuntu-latest image pull in CI. See fail here: https://github.com/viam-modules/video-store/actions/runs/12697709786/job/3539496099

Not doing any JPEG encoding here anyhow so it is mostly just an artifact of the RDK camera dependency.

File: README.md [hunk 6]
```diff
@@ -44,11 +44,6 @@ On the new component panel, copy and paste the following attribute template into
   "storage": {
     "segment_seconds": <int>,
     "size_gb": <int>,
-  },
-  "cam_props": {
-    "width": <int>,
-    "height": <int>,
-    "framerate": <int>
```
[the remark this answers is not in the document]
Comment (seanavery) [reply README.md:51]: In order to avoid stuffing duplicate frames in video segments or skipping frames,  it is nice to accurately match target framerate with the source camera.

File: cam/cam.go (hunk 1 above)
Comment (seanavery) [reply cam/cam.go]: good call out, fixed.

File: cam/cam.go [hunk 7]
```diff
@@ -164,39 +158,16 @@ func newvideostore(
...
 	}
 
 	vs.enc, err = newEncoder(
```
[the remark this answers is not in the document]
Comment (seanavery) [reply cam/cam.go:164]: Good call, changed.

File: cam/encoder.go (hunk 3 above)
Comment (seanavery) [reply cam/encoder.go]: like this -- changed.

File: cam/encoder.go (hunk 4 above)
Comment (seanavery) [reply cam/encoder.go]: Yes thats much cleaner -- changed.

File: cam/cam.go [hunk 8]
```diff
@@ -400,11 +370,19 @@ func (vs *videostore) processFrames(ctx context.Context) {
...
 				vs.logger.Error("failed to encode frame", err)
 				return
 			}
+			if reinit {
+				vs.logger.Info("reinitializing segmenter due to encoder refresh")
+				err = vs.seg.initialize(vs.enc.codecCtx)
+				if err != nil {
+					vs.logger.Error("failed to reinitialize segmenter", err)
+					return
+				}
```
[the remark this answers is not in the document]
Comment (seanavery) [reply cam/cam.go]: Now clearing out the codec context of the encoder in this scenario so that we retry to initialize both again on the next frame.

Somewhat of a hack, but this will allow us to keep atomicity between the two.

File: cam/encoder.go (hunk 2 above)
Comment (seanavery) [reply cam/encoder.go:150]: Yes added a deffered cleanup in the error case.

**Comment by seanavery** [comment]
> Can you please test whether or not there is a significant performance cost to this change?

Did some htop sanity checks on this to verify no change but do not think it is worth diving deeper since this is an O(1) perf cost.

File: cam/cam.go [hunk 9]
```diff
@@ -400,15 +368,26 @@ func (vs *videostore) processFrames(ctx context.Context) {
...
 				vs.logger.Error("failed to encode frame", err)
-				return
+				continue
 			}
-			err = vs.seg.writeEncodedFrame(encoded, pts, dts)
+			if result.frameDimsChanged {
+				vs.logger.Info("reinitializing segmenter due to encoder refresh")
+				err = vs.seg.initialize(vs.enc.codecCtx)
+				if err != nil {
+					vs.logger.Error("failed to reinitialize segmenter", err)
```
Comment (randhid) [inline cam/cam.go:380]: how do you feel about making these errors debug level? Are they edge case enough to warrant that rather than spamming the user?

File: cam/cam.go [hunk 10]
```diff
@@ -164,39 +157,15 @@ func newvideostore(
...
-		}
-	}
-	if newConf.Properties.Width == 0 && newConf.Properties.Height == 0 {
-		return nil, fmt.Errorf("failed to get source camera width and height after %d attempts", numFetchFrameAttempts)
-	}
-
-	if newConf.Properties.Framerate == 0 {
-		newConf.Properties.Framerate = defaultFramerate
+	if newConf.Framerate > 0 {
+		vs.framerate = newConf.Framerate
```
Comment (randhid) [inline cam/cam.go:161]: move the setting of `vs.framerate` above this line - I also think validation should just fail if you set a framerate that's zero or negative - a user has obviously done something wrong for this driver  and we should tell them rather than have it silently build itself.

File: cam/encoder.go [hunk 11]
```diff
@@ -27,91 +27,140 @@ type encoder struct {
...
 // PTS will lag behind actual run time.
-func (e *encoder) encode(frame image.Image) ([]byte, int64, int64, error) {
+func (e *encoder) encode(frame image.Image) (encodeResult, error) {
+	var result encodeResult
+	dy, dx := frame.Bounds().Dy(), frame.Bounds().Dx()
+	if e.codecCtx == nil || dy != int(e.codecCtx.height) || dx != int(e.codecCtx.width) {
+		e.logger.Infof("Initializing encoder with frame dimensions %dx%d", dx, dy)
+		err := e.initialize(dx, dy)
+		if err != nil {
+			return result, err
```
Comment (randhid) [inline cam/encoder.go:152]: is this meant to be an empty result or a nil? might be more readable to make result an explicit `result := encoderResult{}` rather than a `var result encoderResult` because I do not know if that initializes to nil from reading this code.

File: cam/cam.go (hunk 9 above)
Comment (seanavery) [reply cam/cam.go]: Sounds good. We hopefully will not be getting into a state where encoder and/or segmenter are broken and we are spamming in a loop, but changed to debug logs to be safe.

File: cam/encoder.go (hunk 11 above)
Comment (seanavery) [reply cam/encoder.go:152]: Good call to be explicit here with default values for encodeResult so it is easier to read in early return cases. Changed.

**Review by nicksanford — APPROVED** [review]
